codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java

```diff
@@ -467,6 +436,96 @@ public void settingsWriteShouldNotUpdateSettings() throws Exception {
+        reset(future);
+        final ByteBuf data = dummyData();
+        encoder.writeData(ctx, STREAM_ID, data, 0, true, promise);
+        assertEquals(1, payloadCaptor.getAllValues().size());
+        verify(lifecycleManager).closeLocalSide(eq(stream), eq(promise));
+        assertEquals(0, data.refCnt());
+    }
+
+    @Test
+    public void headersWriteShouldHalfCloseStream() throws Exception {
+        doAnswer(new Answer() {
+            @Override
+            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
+                FlowControlled flowControlled = (FlowControlled) invocationOnMock.getArguments()[2];
+                flowControlled.write(Integer.MAX_VALUE);
+                flowControlled.writeComplete();
+                return null;
+            }
+        }).when(remoteFlow).sendFlowControlled(eq(ctx), eq(stream), payloadCaptor.capture());
+        int streamId = 5;
+        when(stream.id()).thenReturn(streamId);
+        when(stream.state()).thenReturn(IDLE);
+        mockFutureAddListener(true);
+        when(local.createStream(eq(streamId))).thenReturn(stream);
+        when(writer.writeHeaders(eq(ctx), eq(streamId), eq(EmptyHttp2Headers.INSTANCE), eq(0),
+                                 eq(DEFAULT_PRIORITY_WEIGHT), eq(false), eq(0), eq(true), eq(promise)))
+                .thenReturn(future);
+        encoder.writeHeaders(ctx, streamId, EmptyHttp2Headers.INSTANCE, 0, true, promise);
+        verify(local).createStream(eq(streamId));
+        verify(stream).open(eq(true));
+        // Trigger the write and mark the promise successful to trigger listeners
+        payloadCaptor.getValue().write(0);
+        promise.trySuccess();
+        verify(lifecycleManager).closeLocalSide(eq(stream), eq(promise));
+    }
+
+    @Test
+    public void headersWriteShouldHalfClosePushStream() throws Exception {
+        doAnswer(new Answer() {
+            @Override
+            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
+                FlowControlled flowControlled = (FlowControlled) invocationOnMock.getArguments()[2];
+                flowControlled.write(Integer.MAX_VALUE);
+                flowControlled.writeComplete();
+                return null;
+            }
+        }).when(remoteFlow).sendFlowControlled(eq(ctx), eq(stream), payloadCaptor.capture());
+
+        mockFutureAddListener(true);
+        when(stream.state()).thenReturn(RESERVED_LOCAL).thenReturn(HALF_CLOSED_LOCAL);
+        when(writer.writeHeaders(eq(ctx), eq(STREAM_ID), eq(EmptyHttp2Headers.INSTANCE), eq(0),
+                                 eq(DEFAULT_PRIORITY_WEIGHT), eq(false), eq(0), eq(true), eq(promise)))
+                .thenReturn(future);
+        encoder.writeHeaders(ctx, STREAM_ID, EmptyHttp2Headers.INSTANCE, 0, true, promise);
+        verify(stream).open(true);
+        // Trigger the write and mark the promise successful to trigger listeners
+        payloadCaptor.getValue().write(0);
```

Comment:
@Scottmitch I didn't write this. I only added the mocking. so no I can't explain it :-).